[11.0][FIX] l10n_es_aeat_sii: float_round error in BaseImponible

eantones opened the pull request:
Example on exempt invoices:

- "Invoice Lines" tab: 
    `{ quantity: 1, price_unit: 82.60, invoice_line_tax_ids: ['IVA Exento Repercutido'] }`
    
- "SII" tab:
    - **"General" tab**

        * Error de envio SII: 1100 | Valor o tipo incorrecto del campo: BaseImponible
        
    - **"Technical" tab**

        * _Último contenido enviado al SII_
        {
            "PeriodoLiquidacion": {
                "Ejercicio": 2019,
                "Periodo": "01"
            },
            "IDFactura": {
                "IDEmisorFactura": {
                    "NIF": "XXXXXXXX"
                },
                "FechaExpedicionFacturaEmisor": "07-01-2019",
                "NumSerieFacturaEmisor": "XXXXXXX"
            },
            "FacturaExpedida": {
                "TipoDesglose": {
                    "DesgloseTipoOperacion": {
                        "PrestacionServicios": {
                            "Sujeta": {
                                "Exenta": {
                                    "DetalleExenta": [
                                        {
                                            "BaseImponible": 82.60000000000001,
                                            "CausaExencion": "E1"
                                        }
                                    ]
                                }
                            }
                        }
                    }
                },
                "DescripcionOperacion": "/",
                "ImporteTotal": 82.6,
                "TipoFactura": "F2",
                "ClaveRegimenEspecialOTrascendencia": "01"
            }
        }

        * _Retorno SII_ 
        {
            'CSV': None,
            'DatosPresentacion': None,
            'Cabecera': {
                'IDVersionSii': '1.1',
                'Titular': {
                    'NombreRazon': 'XXXXXXXXXXXXXXXXXXXX',
                    'NIFRepresentante': None,
                    'NIF': 'XXXXXXXXXXX'
                },
                'TipoComunicacion': 'A0'
            },
            'EstadoEnvio': 'Incorrecto',
            'RespuestaLinea': [
                {
                    'IDFactura': {
                        'IDEmisorFactura': {
                            'NIF': 'XXXXXXXXXXX'
                        },
                        'NumSerieFacturaEmisor': 'XXXXXXXXXXXXX',
                        'NumSerieFacturaEmisorResumenFin': None,
                        'FechaExpedicionFacturaEmisor': '07-01-2019'
                    },
                    'RefExterna': None,
                    'EstadoRegistro': 'Incorrecto',
                    'CodigoErrorRegistro': 1100,
                    'DescripcionErrorRegistro': 'Valor o tipo incorrecto del campo: BaseImponible',
                    'CSV': None,
                    'RegistroDuplicado': None
                }
            ]
        }

pedrobaeza commented:
@eantones he visto que hay otros 3 sitios en donde se utiliza `base_company` sin redondear. ¿Podrías añadir el redondeo también ahí?

eantones committed:
[FIX] l10n_es_aeat_sii: float_round error in BaseImponible

Example on exempt invoices:

- "Invoice Lines" tab:
    * { quantity: 1, price_unit: 82.60, invoice_line_tax_ids: ['IVA Exento Repercutido'] }

- "SII" tab:
    - "General" tab:

        * Error de envio SII: 1100 | Valor o tipo incorrecto del campo: BaseImponible

    - "Technical" tab:

        * Último contenido enviado al SII:
        {
            "PeriodoLiquidacion": {
                "Ejercicio": 2019,
                "Periodo": "01"
            },
            "IDFactura": {
                "IDEmisorFactura": {
                    "NIF": "XXXXXXXX"
                },
                "FechaExpedicionFacturaEmisor": "07-01-2019",
                "NumSerieFacturaEmisor": "XXXXXXX"
            },
            "FacturaExpedida": {
                "TipoDesglose": {
                    "DesgloseTipoOperacion": {
                        "PrestacionServicios": {
                            "Sujeta": {
                                "Exenta": {
                                    "DetalleExenta": [
                                        {
                                            "BaseImponible": 82.60000000000001,
                                            "CausaExencion": "E1"
                                        }
                                    ]
                                }
                            }
                        }
                    }
                },
                "DescripcionOperacion": "/",
                "ImporteTotal": 82.6,
                "TipoFactura": "F2",
                "ClaveRegimenEspecialOTrascendencia": "01"
            }
        }

        * Retorno SII
        {
            'CSV': None,
            'DatosPresentacion': None,
            'Cabecera': {
                'IDVersionSii': '1.1',
                'Titular': {
                    'NombreRazon': 'XXXXXXXXXXXXXXXXXXXX',
                    'NIFRepresentante': None,
                    'NIF': 'XXXXXXXXXXX'
                },
                'TipoComunicacion': 'A0'
            },
            'EstadoEnvio': 'Incorrecto',
            'RespuestaLinea': [
                {
                    'IDFactura': {
                        'IDEmisorFactura': {
                            'NIF': 'XXXXXXXXXXX'
                        },
                        'NumSerieFacturaEmisor': 'XXXXXXXXXXXXX',
                        'NumSerieFacturaEmisorResumenFin': None,
                        'FechaExpedicionFacturaEmisor': '07-01-2019'
                    },
                    'RefExterna': None,
                    'EstadoRegistro': 'Incorrecto',
                    'CodigoErrorRegistro': 1100,
                    'DescripcionErrorRegistro': 'Valor o tipo incorrecto del campo: BaseImponible',
                    'CSV': None,
                    'RegistroDuplicado': None
                }
            ]
        }

eantones commented:
@pedrobaeza Hecho!!

Gracias

SoniaViciana commented:
Buenos días,

Revisando la codificación que Odoo envía al SII, me encuentro con posibles errores en la generación de facturas rectificativas. 
Odoo da tres opciones a la hora de rectificar una factura:

![image](https://user-images.githubusercontent.com/46522913/50895766-4ff08480-1407-11e9-8798-5ab678911bed.png)


1. **Crear un borrador de rectificativa**: Genera un borrador de factura para editar. Opción adecuada para realizar una factura rectificativa por **diferencias**. La codificación que se envía por defecto es:

- A0,**R4**, I :x:

Revisando el descriptivo de las distintas rectificativas, se entiende que lo correcto sería que por defecto se emitiera con **R1**, puesto que R4 se entiende que es para aquellos casos en los que la rectificación no se ajuste a la Ley del IVA, lo que podría generar que la AEAT nos solicite justificar esas rectificaciones fuera de la legislación, a posteriori. Sin embargo R1, sí que se ajusta a la gran mayoría de rectificaciones que hacen los comercios en el apartado 1 Y 2 del Artículo 80:

> Uno. La base imponible determinada con arreglo a lo dispuesto en los artículos 78 y 79 anteriores se reducirá en las cuantías siguientes:
> 1.º El importe de los **envases y embalajes** susceptibles de reutilización que hayan sido **objeto de devolución**.
> 2.º Los **descuentos y bonificaciones otorgados con posterioridad al momento en que la operación se haya realizado** siempre que sean debidamente justificados.
> Dos. Cuando por resolución firme, judicial o administrativa o con arreglo a Derecho o a los **usos de comercio queden sin efecto total o parcialmente las operaciones gravadas o se altere el precio después del momento en que la operación se haya efectuado**, la base imponible se modificará en la cuantía correspondiente.
__________________________________________________________________________________

2. **Cancelar: crear rectificativa y conciliar**: Esta opción abona o anula totalmente la factura original. La codificación que se envía al SII por defecto es:

- **A0, R4,I**  :x:

Según se define, sería más adecuado tratarlo como una "anulación" de la factura original:

> Cuando se emite una factura negativa siempre se debe justificar los motivos de la anulación. Los supuestos para la emisión de una factura negativa pueden ser, la cancelación de un producto que ya se ha cobrado, pero no se ha recibido o la cancelación de un servicio ya abonado pero que nunca se ha producido, o la **devolución de un producto**, etc.

En tal caso, la codificación correcta sería:

- **A0, F1 (base imponible + cuota impuesto en negativo)** :white_check_mark:
__________________________________________________________________________________

3. **Modificar: emitir rectificativa, conciliar y crear una nueva factura en borrador** Anula la factura original, y genera una factura positiva en borrador. Opción adecuada para realizar una factura rectificativa por **sustitución**. La codificación que se envía por defecto es:

- ANULACIÓN: **A0, R4, I** :x:
- FACTURA BORRADOR (positiva): **A0, F1** :x:

Sin embargo, lo más adecuado sería que se enviara la codificación por defecto estipulada por la AEAT para las facturas rectificactivas por sustitución:

- ANULACIÓN: **A0, F1 (base imponible + cuota impuesto en negativo)** :white_check_mark:
- FACTURA BORRADOR (positiva): **A0, R1, S** :white_check_mark:

Aplicando esta regla, sería aconsejable eliminar la opción seleccionable "Tipo rectificativa SII":

![image](https://user-images.githubusercontent.com/46522913/50895813-7adad880-1407-11e9-9e39-2c45e3c9fd06.png)


@liebana

eantones commented:
Hola @pedrobaeza,
he detectado otros sitios dónde genera el mismo error de BaseImponible, hago un nuevo PR o utilizo este mismo?
Gracias

pedrobaeza commented:
Éste ya está fusionado, así que tendrás que hacer uno nuevo. No sé dónde dices que pueden quedar, pero ya revisaré sobre el PR.

eantones commented:
Por ejemplo, en estas líneas https://github.com/OCA/l10n-spain/blob/fabc3a39eda6c595045906a3a612a18a6d902f9f/l10n_es_aeat_sii/models/account_invoice.py#L537-L538 la suma (+=) genera otro error de coma flotante, debugando me encuentro con la siguiente suma ya después de haber redondeado: 170.61 + 107.35, si lo haces en la consola de Python ya verás que te devuelve 277.96000000000004. En algún momento se debería utilizar el paquete de Python "decimal" para eliminar completamente el clásico problema inherente en el coma flotante y que hace que la ejecución del código no se estable.
Ok hago un nuevo PR con toda esta información, muchas gracias!
